### PR TITLE
nginx: Enable HSTS when forceSSL is on

### PR DESCRIPTION
**Pull request by kirelagin**

### CODE REVIEW

File: nixos/modules/services/web-servers/nginx/default.nix [hunk 1]
```diff
@@ -170,6 +170,9 @@ let
             ssl_certificate ${vhost.sslCertificate};
             ssl_certificate_key ${vhost.sslCertificateKey};
           ''}
+          ${optionalString vhost.forceSSL ''
+            add_header Strict-Transport-Security "max-age=31536000; includeSubDomains" always;
```
Comment (Mic92) [inline nixos/modules/services/web-servers/nginx/default.nix]: This should be definitely mentioned in both documention for `forceSSL` and nixos release changelog. There might be also the use case for people wanting a redirect, but no `HSTS` (yet), when testing.

Comment (kirelagin) [inline nixos/modules/services/web-servers/nginx/default.nix]: Yes, I agree that some people might want that. At the same time my point is that, if not for testing, compulsory redirect without HSTS is useless in terms of security.

How would you suggest to implement that? Have a completely separate config option? Have a separate config option, enabled by default by `forceSSL` and mentioned in its documentation? Other ideas?

**Review by fpletz — CHANGES_REQUESTED** [review]
This change is very dangerous. We thought about adding HSTS when we implemented `forceSSL` but did not because it could break websites and should therefore be a conscious user choice. Including `includeSubDomains` by default would also probably break other sites unexpectedly.

The only option I think is feasible to create a separate option to enable HSTS and maybe make `max-age` and `includeSubDomains` configurable.

**Comment by kirelagin** [comment]
> The only option I think is feasible to create a separate option to enable HSTS and maybe make max-age and includeSubDomains configurable.

Ok, in this case I’ll add a separate option and mention it in the documentation for `forceSSL`.

**Comment by domenkozar** [comment]
It's very easy to lock users out of the website, I would never enable HSTS by default, but we can make it optional.

**Comment by kirelagin** [comment]
> It's very easy to lock users out of the website

Well, personally I do not think that HSTS is _that_ scary, it does not pin a certificate and obtaining a valid certificate nowadays is incredibly easy.

Anyway, I’m thinking about something like:

~~~~
virtualHosts."example.com" = {
  sslStrict = {
    enable = false;
    maxAge = 31536000;
    includeSubDomains = true;
  };
};
~~~~

I am also inclined towards always sending the header, but setting `max-age=0` if `enable` is `false`. That way changing `enable` to `false` will actually _disable_ HSTS, not just stop sending the header.

**Comment by kirelagin** [comment]
In case someone wonders, I still hope to get to this some time soon.

**Comment by JohnAZoidberg** [comment]
> Anyway, I’m thinking about something like:
> 
> ```
> virtualHosts."example.com" = {
>   sslStrict = {
>     enable = false;
>     maxAge = 31536000;
>     includeSubDomains = true;
>   };
> };
> ```

Yeah, that looks good!
We probably shouldn't, like the others mentioned, turn it on by default. But I think we can make the case that `recommendedTlsSettings` turns it on.
Oh and `includeSubDomains` should be false by default.
We should also probably set the default `maxAge` to a low number (like an hour) and also recommend in the docs to set it higher. (Which `recommendedTlsSettings` could do)

> I am also inclined towards always sending the header, but setting `max-age=0` if `enable` is `false`. That way changing `enable` to `false` will actually _disable_ HSTS, not just stop sending the header.

Hmm, no I think people wouldn't expect that. You don't want to mess with the entire domain if not explicitly opted in.

**Comment by solson** [comment]
Given the complications around disabling HSTS, I wouldn't like to see it enabled implicitly (like by `recommendedTlsSettings`), but I would love to have an option just like the suggested `sslStrict` for doing it manually, but easily.

(Incidentally, are there any functions in nixpkgs for dealing with units of time? It would be cool to have `maxAge = lib.daysToSeconds 365` (or something more structured) instead of `31536000`.)

Is anyone else working on this? If not, I will make a PR for `sslStrict` when I find the time.

## About sending `max-age=0` by default

I am also in favor of not sending the header at all if `enable = false` (i.e. by default). I'm not an expert on this, but I believe in some imperfect configurations, some pages on a domain might provide `max-age=31536000` while other pages (perhaps served by a different server) serve no HSTS header at all - the browser will still force those non-HSTS pages to load as HTTPS as long as it has seen any HSTS page on the domain previously within the `max-age` window. But if those non-HSTS pages were to send `max-age=0` by default, without the sysadmin's knowledge, HSTS would effectively be getting turned on and off depending on which pages are visited.

tl;dr: We should only send `max-age=0` when the sysadmin specifically chooses to do so, because I think it's a non-trivial decision and we can't be certain the default won't cause problems.

**Comment by Mic92** [comment]
I would prefer this to be a seperate option rather than beeing included by default.
